substrate/frame/opf/src/functions.rs

```diff
@@ -0,0 +1,360 @@
+	) -> DispatchResult {
+		let origin = T::RuntimeOrigin::from(RawOrigin::Signed(voter_id.clone()));
+		if !ProjectFunds::<T>::contains_key(&project) {
+			let bounded = BoundedVec::<BalanceOf<T>, ConstU32<2>>::try_from(vec![
+				BalanceOf::<T>::zero(),
+				BalanceOf::<T>::zero(),
+			])
+			.expect("It works");
+			ProjectFunds::<T>::insert(&project, bounded);
+		}
+
+		let infos = WhiteListedProjectAccounts::<T>::get(project.clone())
+			.ok_or(Error::<T>::NoProjectAvailable)?;
+		let ref_index = infos.index;
+
+		let conviction_fund =
+			Self::conviction_amount(amount, conviction).ok_or("Invalid conviction")?;
+
+		// Create vote infos and store/adjust them
+		let round_number = NextVotingRoundNumber::<T>::get().saturating_sub(1);
+		let mut round = VotingRounds::<T>::get(round_number).ok_or(Error::<T>::NoRoundFound)?;
+		if is_fund {
+			round.total_positive_votes_amount =
+				round.total_positive_votes_amount.saturating_add(conviction_fund);
+		} else {
+			round.total_negative_votes_amount =
+				round.total_negative_votes_amount.saturating_add(conviction_fund);
+		}
+
+		VotingRounds::<T>::mutate(round_number, |val| {
+			*val = Some(round.clone());
+		});
+
+		let mut new_vote = VoteInfo {
+			amount,
+			round: round.clone(),
+			is_fund,
+			conviction,
+			funds_unlock_block: round.round_ending_block,
+		};
+
+		// Update Funds unlock block according to the selected conviction
+		new_vote.funds_unlock();
+		if Votes::<T>::contains_key(&project, &voter_id) {
+			let old_vote = Votes::<T>::get(&project, &voter_id).ok_or(Error::<T>::NoVoteData)?;
+			let old_amount = old_vote.amount;
+			let old_conviction = old_vote.conviction;
+			let old_conviction_amount =
+				Self::conviction_amount(old_amount, old_conviction).ok_or("Invalid conviction")?;
+			ProjectFunds::<T>::mutate(&project, |val| {
+				let mut val0 = val.clone().into_inner();
+				if is_fund {
+					val0[0] = val0[0 as usize]
+						.saturating_add(conviction_fund)
+						.saturating_sub(old_conviction_amount);
+				} else {
+					val0[1] = val0[1 as usize]
+						.saturating_add(conviction_fund)
+						.saturating_sub(old_conviction_amount);
+				}
+				*val = BoundedVec::<BalanceOf<T>, ConstU32<2>>::try_from(val0).expect("It works");
+			});
+
+			Votes::<T>::mutate(&project, &voter_id, |value| {
+				*value = Some(new_vote);
+			});
+
+			// Adjust locked amount
+			let total_hold = T::NativeBalance::total_balance_on_hold(&voter_id);
+			let new_hold = total_hold.saturating_sub(old_amount).saturating_add(amount);
+			T::NativeBalance::set_on_hold(&HoldReason::FundsReserved.into(), &voter_id, new_hold)?;
+
+			// Remove previous vote from Referendum
+			Democracy::Pallet::<T>::remove_vote(origin, ref_index)?;
+		} else {
+			Votes::<T>::insert(&project, &voter_id, new_vote);
+			ProjectFunds::<T>::mutate(&project, |val| {
+				let mut val0 = val.clone().into_inner();
+				if is_fund {
+					val0[0] = val0[0 as usize].saturating_add(conviction_fund);
```

Comment:
Please use a struct and not just a vec with two values. This makes it much more complicated to read the code.